Src/library/net/Clib/network.c

```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 #ifdef EIF_WINDOWS
 	int arg;
 	asize = sizeof(arg);
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, (char *) &arg, &asize));
+	return RTU64C(1000000) * arg; /* 1 millisecond = 1000000 nanoseconds */
 #else
 	struct timeval tv;
 	asize = sizeof(struct timeval);
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, &tv, &asize));
-	return (EIF_INTEGER) tv.tv_sec;
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, &tv, &asize));
+	/* 1 second = 1000000000 nanoseconds, and 1 microsecond = 1000 nanoseconds */
+	return RTU64C(1000000000) * tv.tv_sec + RTU64C(1000) * tv.tv_usec;
 #endif
 }
 
-void c_set_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_seconds)
-	/* set socket SO_RCVTIMEO option to recv_timeout_seconds */
+void c_set_sock_opt_timeo(EIF_INTEGER fd, EIF_INTEGER level, int optname, EIF_NATURAL_64 timeout_nano_seconds)
+	/* set socket `optname` option to timeout_nano_seconds */
 {
 #ifdef EIF_WINDOWS
-	int arg = (int) 1000 * recv_timeout_seconds; /* Timeout in milliseconds */
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_RCVTIMEO, (char *) &arg, sizeof(arg)));
+	int arg = (int)(timeout_nano_seconds / 1000000); /* Timeout in milliseconds: 1000000 nanoseconds =  1 millisecond */
```

Comment:
Let's put an assert about the maximum value of timeout_nano_seconds to ensure that there will never be an overflow. Put the cooresponding require on the Eiffel side.

Reply:
Why didn't you do the assert path since we already have the protection on the Eiffel side now?